### PR TITLE
[API] Add product endpoints

### PR DESCRIPTION
**Pull request by mpysiak**
| Q               | A
|-----------------|-----
| Branch?         | api-platform-3
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | n/a
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php [hunk 1]
```diff
@@ -41,9 +41,6 @@
  * The fix has been merged in PR #6367 (https://github.com/api-platform/core/pull/6367), and we are waiting for the new version of ApiPlatform to be released.
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: Given this comment, it's possible we can just remove this whole class.

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php [hunk 2]
```diff
@@ -1,9 +1,9 @@
 <?php
 
 /*
- * This file is part of the API Platform project.
+ * This file is part of the Sylius package.
  *
- * (c) Kévin Dunglas <dunglas@gmail.com>
+ * (c) Sylius Sp. z o.o.
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: To be reverted, it's not our code

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php:1]: I don't get why was this class touched at all in this PR, nothing's changed, just some ecs shenanigans

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    <resource class="Sylius\Component\Core\Model\Product" shortName="Product">
+        <validationContext>
+            <values>
+                <value name="groups">
+                    <values>
+                        <value>sylius</value>
+                    </values>
+                </value>
+            </values>
+        </validationContext>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml:28]: It's an overall comment, so I'll just note it here:
We should refrain from setting global validation groups for a whole resource, it makes changing/overwriting, or even adding new endpoints for it tedious.

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 4]
```diff
@@ -776,6 +776,11 @@ private function hasProductOptionWithNameAndValue(string $expectedOptionName, st
 
         foreach ($productVariants as $productVariant) {
             foreach ($productVariant['optionValues'] as $optionValueIri) {
+
+                if (is_array($optionValueIri)) {
+                    continue;
+                }
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php]: Why?

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php [hunk 5]
```diff
@@ -1,9 +1,9 @@
 <?php
 
 /*
- * This file is part of the API Platform project.
```
Comment (TheMilek) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: I would leave it as it is, as it was added like that on purpose 🤔  @Wojdylak

**Review by NoResponseMate — CHANGES_REQUESTED** [review]
Lotsa unrelated changes, but removing them atm would be wasteful so let's just leave it as is.
On a different note, we're missing Product's contract tests.

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/ProductReview.xml
Comment (TheMilek) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/ProductReview.xml:1]: Was it actually necessary to add it to upgrade the product resource?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Taxon.xml
Comment (TheMilek) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Taxon.xml:1]: Same here, I would be after sticking only to the dedicated resource and if IRI needed from other ones, just adding some item get endpoint to avoid such a big PR's

File: src/Sylius/Bundle/ApiBundle/StateProvider/Shop/OrderItemAdjustmentsProvider.php [hunk 6]
```diff
@@ -25,7 +25,7 @@ public function __construct(
     ) {
     }
 
-    public function provide(Operation $operation, array $uriVariables = [], array $context = []): Collection|array
```
Comment (TheMilek) [inline src/Sylius/Bundle/ApiBundle/StateProvider/Shop/OrderItemAdjustmentsProvider.php:28]: we should use ecs on related to the PR changes to prevent it from growing too big, put it in separate PR next time 🤔

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php (hunk 1 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: If we bump APIP to version 3.3.5, we can remove this class.

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php (hunk 5 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: https://github.com/Sylius/Sylius/blob/api-platform-3/src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php#L40-L41

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php (hunk 4 above)
Comment (mpysiak) [reply src/Sylius/Behat/Context/Api/Shop/ProductContext.php]: It was some test fix for `viewing_products` scenario. Reverted it since `view_products` scenario is not enabled after all

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php (hunk 1 above)
Comment (mpysiak) [reply src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: https://github.com/Sylius/Sylius/actions/runs/9383182395/job/25836332303#step:9:288

Removed the class

Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: I would prefer to lock the minimum version to 3.3.4 to ensure that the adjustments will work.

Comment (NoResponseMate) [reply src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: From what I see the newest is 3.3.5, we can just set that as the minimum

**Review by NoResponseMate — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductImagesContext.php [hunk 7]
```diff
@@ -18,22 +18,22 @@
...
-final class ManagingProductImagesContext implements Context
+final readonly class ManagingProductImagesContext implements Context
 {
     public function __construct(
         private ApiClientInterface $client,
         private ResponseCheckerInterface $responseChecker,
         private SharedStorageInterface $sharedStorage,
         private \ArrayAccess $minkParameters,
-        private SectionAwareIriConverter $sectionAwareIriConverter,
+        private IriConverterInterface $sectionAwareIriConverter,
```
Comment (Wojdylak) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductImagesContext.php]:
```suggestion
        private IriConverterInterface $iriConventer,
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingTaxonsContext.php [hunk 8]
```diff
@@ -18,21 +18,21 @@
...
-final class ManagingTaxonsContext implements Context
+final readonly class ManagingTaxonsContext implements Context
 {
     use ValidationTrait;
 
     public function __construct(
         private ApiClientInterface $client,
         private ResponseCheckerInterface $responseChecker,
-        private SectionAwareIriConverterInterface $sectionAwareIriConverter,
+        private IriConverterInterface $sectionAwareIriConverter,
```
Comment (Wojdylak) [inline src/Sylius/Behat/Context/Api/Admin/ManagingTaxonsContext.php]:
```suggestion
        private IriConverterInterface $iriConverter,
```

File: src/Sylius/Behat/Resources/config/services/contexts/api/admin.xml [hunk 9]
```diff
@@ -62,12 +62,12 @@
             <argument type="service" id="behat.mink.parameters" />
         </service>
 
-<!--        <service id="Sylius\Behat\Context\Api\Admin\ManagingProductTaxonsContext">-->
-<!--            <argument type="service" id="sylius.behat.api_platform_client.admin" />-->
-<!--            <argument type="service" id="api_platform.iri_converter" />-->
-<!--            <argument type="service" id="Sylius\Behat\Client\ResponseCheckerInterface" />-->
-<!--            <argument type="service" id="sylius.behat.shared_storage" />-->
-<!--        </service>-->
+        <service id="Sylius\Behat\Context\Api\Admin\ManagingProductTaxonsContext">
```
Comment (Wojdylak) [inline src/Sylius/Behat/Resources/config/services/contexts/api/admin.xml]:
```suggestion
        <service id="sylius.behat.context.api.admin.managing_product_taxons" class="Sylius\Behat\Context\Api\Admin\ManagingProductTaxonsContext">
```
Since version 2.0, we have started naming services in this way. More information can be found [here](https://github.com/Sylius/Sylius/pull/15883)

File: src/Sylius/Behat/Resources/config/services/contexts/api/shop.xml [hunk 10]
```diff
@@ -190,11 +190,11 @@
 <!--            <argument type="service" id="sylius.behat.api_platform_client.shop" />-->
 <!--        </service>-->
 
-<!--        <service id="Sylius\Behat\Context\Api\Shop\TaxonContext">-->
-<!--            <argument type="service" id="sylius.behat.api_platform_client.shop" />-->
-<!--            <argument type="service" id="Sylius\Behat\Client\ResponseCheckerInterface" />-->
-<!--            <argument type="service" id="api_platform.iri_converter" />-->
-<!--            <argument type="service" id="doctrine.orm.entity_manager" />-->
-<!--        </service>-->
+        <service id="Sylius\Behat\Context\Api\Shop\TaxonContext">
```
Comment (Wojdylak) [inline src/Sylius/Behat/Resources/config/services/contexts/api/shop.xml]:
```suggestion
        <service id="..." class="Sylius\Behat\Context\Api\Shop\TaxonContext">
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/EnabledProductInProductAssociationExtension.php [hunk 11]
```diff
@@ -11,30 +11,47 @@
...
+        $this->modifyQueryBuilder($queryBuilder, $queryNameGenerator, $resourceClass);
+    }
+
+    private function modifyQueryBuilder(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
     ): void {
         if (!is_a($resourceClass, ProductInterface::class, true)) {
             return;
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/EnabledProductInProductAssociationExtension.php:58]: Below is a check to verify if the user is an admin. It would be ideal to use the same method in all new classes by using `sectionProvider`
```
if (!$this->sectionProvider->getSection() instanceof AdminApiSection) {
    return;
}
```

Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/EnabledProductInProductAssociationExtension.php:58]: Below is a check to verify if the user is an admin. It would be ideal to use the same method in all new classes by using `sectionProvider`
```
if (!$this->sectionProvider->getSection() instanceof AdminApiSection) {
    return;
}
```

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Routing/IriConverter.php [hunk 12]
```diff
@@ -28,22 +29,23 @@ public function __construct(
...
     public function getIriFromResource(
         object|string $resource,
         int $referenceType = UrlGeneratorInterface::ABS_PATH,
-        Operation $operation = null,
+        ?Operation $operation = null,
         array $context = [],
     ): ?string {
         $resourceClass = $context['force_resource_class'] ?? (\is_string($resource) ? $resource : $this->getObjectClass($resource));
-        $pathPrefix = $this->pathPrefixProvider->getPathPrefix($context['request_uri'] ?? '');
+        $pathPrefix = $this->pathPrefixProvider->getPathPrefix($context['request_uri'] ?? $this->router->getContext()->getPathInfo());
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Routing/IriConverter.php:48]: Out of curiosity, could you explain why you added this?

File: src/Sylius/Bundle/ApiBundle/Resolver/PathPrefixBasedOperationResolver.php [hunk 13]
```diff
@@ -37,6 +38,10 @@ public function resolve(string $resourceClass, ?string $pathPrefix, ?Operation $
                     continue;
                 }
 
+                if (!$resourceOperation instanceof Get) {
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resolver/PathPrefixBasedOperationResolver.php:41]: Where are the tests? :stuck_out_tongue:

File: src/Sylius/Bundle/ApiBundle/Resources/config/integrations/swagger.xml [hunk 14]
```diff
@@ -43,9 +43,9 @@
 <!--            <tag name="sylius.open_api.modifier" />-->
 <!--        </service>-->
 
-<!--        <service id="Sylius\Bundle\ApiBundle\OpenApi\Documentation\ProductDocumentationModifier">-->
-<!--            <tag name="sylius.open_api.modifier" />-->
-<!--        </service>-->
+        <service id="Sylius\Bundle\ApiBundle\OpenApi\Documentation\ProductDocumentationModifier">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/integrations/swagger.xml]:
```suggestion
        <service id"..." class="Sylius\Bundle\ApiBundle\OpenApi\Documentation\ProductDocumentationModifier">
```

File: src/Sylius/Bundle/ApiBundle/StateProcessor/ProductProcessor.php [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    public function __construct(
+        private ProcessorInterface $persistProcessor,
+        private ProcessorInterface $removeProcessor,
+    ) {
+    }
+
+    /**
+     * @param ProductInterface $data
+     */
+    public function process(mixed $data, Operation $operation, array $uriVariables = [], array $context = [])
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/ProductProcessor.php]: ProductProcessorSpec :stuck_out_tongue:

File: src/Sylius/Bundle/ApiBundle/StateProvider/ProductAttributesProvider.php [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+use ApiPlatform\Doctrine\Orm\Util\QueryNameGenerator;
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\State\ProviderInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Locale\Provider\LocaleProviderInterface;
+use Sylius\Component\Product\Model\ProductAttributeValue;
+use Sylius\Component\Product\Repository\ProductAttributeValueRepositoryInterface;
+
+/**
+ * @experimental
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/StateProvider/ProductAttributesProvider.php]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/StateProvider/ProductAttributesProvider.php [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @implements ProviderInterface<ProductAttributeValue>
+ */
+final readonly class ProductAttributesProvider implements ProviderInterface
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/StateProvider/ProductAttributesProvider.php]: ProductAttributesProviderSpec

File: src/Sylius/Bundle/ApiBundle/composer.json [hunk 18]
```diff
@@ -25,7 +25,7 @@
     "require": {
         "php": "^8.2",
         "doctrine/dbal": "^3.0",
-        "api-platform/core": "^3.1",
+        "api-platform/core": "^3.3",
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/composer.json]:
```suggestion
        "api-platform/core": "^3.3.5",
```
or
```suggestion
        "api-platform/core": "~3.3.5",
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml [hunk 19]
```diff
@@ -47,44 +47,44 @@
...
 <!--        </service>-->
 
-<!--        <service id="sylius.api.product_association_filter" parent="api_platform.doctrine.orm.search_filter" public="true">-->
-<!--            <argument type="collection">-->
-<!--                <argument key="owner.code">partial</argument>-->
-<!--                <argument key="type.code">partial</argument>-->
-<!--            </argument>-->
-<!--            <tag name="api_platform.filter" />-->
-<!--        </service>-->
+        <service id="sylius.api.product_association_filter" parent="api_platform.doctrine.orm.search_filter" public="true">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/filters.xml:50]: There was a plan to start naming the filters in a new way since version 2.0, but I don't know if it is still valid. By using the prefix `sylius.api.order_filter` and making them only for one field.

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php:1]: This one could be removed in a separate PR

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml [hunk 20]
```diff
@@ -42,6 +42,26 @@
...
+                <normalizationContext>
+                    <values>
+                        <value name="groups">
+                            <values>
+                                <value>sylius:admin:admin_user:show</value>
+                            </values>
+                        </value>
+                    </values>
+                </normalizationContext>
+            </operation>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml:64]: Why do you need a PUT for the admin user in this PR? 😅

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+                            </values>
+                        </value>
+                    </values>
+                </normalizationContext>
+                <filters>
+                    <filter>sylius.api.product_name_filter</filter>
+                    <filter>sylius.api.product_order_filter</filter>
+                    <filter>sylius.api.product_taxon_code_filter</filter>
+                    <filter>Sylius\Bundle\ApiBundle\Filter\Doctrine\ChannelsAwareChannelFilter</filter>
+                    <filter>Sylius\Bundle\ApiBundle\Filter\Doctrine\TranslationOrderNameAndLocaleFilter</filter>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml]: It would be good to make the names of the services here consistent and not use FQCN

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+                            </values>
+                        </value>
+                    </values>
+                </openapiContext>
+            </operation>
+
+            <operation
+                class="ApiPlatform\Metadata\Get"
+                uriTemplate="/shop/products-by-slug/{slug}"
+                controller="Sylius\Bundle\ApiBundle\Controller\GetProductBySlugAction"
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml:206]: Is controller still the best option in APIP to get a product by slug? 🤔

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml [hunk 23]
```diff
@@ -0,0 +1,328 @@
...
+                    </values>
+                </normalizationContext>
+            </operation>
+
+            <operation class="ApiPlatform\Metadata\Delete" uriTemplate="/admin/products/{code}" />
+
+        </operations>
+    </resource>
+
+    <resource class="Sylius\Component\Core\Model\Product" uriTemplate="/admin/products/{code}/images">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml]:
```suggestion
    <resource class="Sylius\Component\Core\Model\ProductImage" uriTemplate="/admin/products/{code}/images">
```
and move this resource definition to `ProductImage.xml`? 🤔

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Routing/IriConverter.php (hunk 12 above)
Comment (mpysiak) [reply src/Sylius/Bundle/ApiBundle/ApiPlatform/Routing/IriConverter.php:48]: I've got few cases (for shop endpoint) where in the response iri was from admin section. After debugging I've added this, because sometimes `$context` was an empty array

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+                                </value>
+                            </values>
+                        </value>
+                    </values>
+                </openapiContext>
+            </operation>
+        </operations>
+    </resource>
+
+    <resource class="Sylius\Component\Product\Model\ProductAttributeValue" uriTemplate="/shop/products/{code}/attributes">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml]: Shouldn't it be in `ProductAttributeValue.xml` file? 🤔

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml [hunk 25]
```diff
@@ -0,0 +1,328 @@
...
+                            <values>
+                                <value>sylius:shop:product_attribute_value:show</value>
+                            </values>
+                        </value>
+                    </values>
+                </normalizationContext>
+            </operation>
+        </operations>
+    </resource>
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/Product.xml]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml [hunk 26]
```diff
@@ -42,5 +42,11 @@
             <argument type="service" id="sylius.repository.avatar_image" />
             <tag name="api_platform.state_processor" />
         </service>
+
+        <service id="Sylius\Bundle\ApiBundle\StateProcessor\ProductProcessor">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml]: Make the service id consistent with others in this file

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml [hunk 27]
```diff
@@ -49,5 +49,14 @@
             <argument type="service" id="Sylius\Bundle\CoreBundle\SectionResolver\SectionProviderInterface" />
             <tag name="api_platform.state_provider" priority="10" />
         </service>
+
+        <service id="Sylius\Bundle\ApiBundle\StateProvider\ProductAttributesProvider">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml]: Do not use FQCN for a service id

File: tests/Api/Admin/ProductsTest.php
Comment (GSadee) [inline tests/Api/Admin/ProductsTest.php:1]: Remove tests that have been enabled from the `disabledTests` directory

File: tests/Api/Shop/ProductsTest.php
Comment (GSadee) [inline tests/Api/Shop/ProductsTest.php:1]: Remove tests that have been enabled from the `disabledTests` directory

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml (hunk 20 above)
Comment (mpysiak) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml:64]: That was tricky 😄  Some `managing_product` scenarios require changing admin locales.

Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml:64]: These scenarios should probably wait for implementing admin user endpoints 😅, but as it was already done, it's ok 👌🏻

File: features/shop/product/viewing_products/viewing_products_with_multiple_variants_sorted_by_price.feature [hunk 28]
```diff
@@ -22,15 +22,15 @@ Feature: Viewing products with multiple variants sorted by price
         And the product "Johnnie Walker" has "Johnnie Walker Black Label" variant priced at "$25.00"
         And the "Johnnie Walker White Label" product variant is disabled
 
-    @ui @api
+    @todo-api @ui
     Scenario: Sorting products by price of their first variant with ascending order
```
Comment (GSadee) [inline features/shop/product/viewing_products/viewing_products_with_multiple_variants_sorted_by_price.feature:26]: if I am thinking correctly, these scenarios of sorting products should be achievable 🤔 However, considering the size of the PR, let's already cover this separately

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+/**
+ * This class is being overridden due to an issue with the clone method in the Adjustment class.
+ * The fix has been merged in PR #6367 (https://github.com/api-platform/core/pull/6367), and we are waiting for the new version of ApiPlatform to be released.
+ *
+ * Bridges persistence and the API system.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ * @author Baptiste Meyer <baptiste.meyer@gmail.com>
+ */
+final class WriteListener
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Symfony/EventListener/WriteListener.php]: Sth has gone wrong with the rebase probably and this one should be already removed but it is added again 🙈

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml [hunk 30]
```diff
@@ -42,6 +42,26 @@
                     </values>
                 </normalizationContext>
             </operation>
+            <operation class="ApiPlatform\Metadata\Put" uriTemplate="admin/administrators/{id}.{_format}">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/AdminUser.xml:45]:
```suggestion
            <operation class="ApiPlatform\Metadata\Put" uriTemplate="admin/administrators/{id}">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/ProductImage.xml [hunk 31]
```diff
@@ -0,0 +1,288 @@
...
+                    </values>
+                </normalizationContext>
+            </operation>
+
+            <operation class="ApiPlatform\Metadata\Delete" uriTemplate="/admin/product-images/{id}" />
+
+        </operations>
+    </resource>
+
+    <resource class="Sylius\Component\Core\Model\ProductImage" uriTemplate="/admin/products/{code}/images">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/ProductImage.xml:259]: Why did you add this endpoint both here and in `Product.xml` if I see correctly?

File: src/Sylius/Bundle/ApiBundle/StateProcessor/Delete/ProductProcessor.php [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+use ApiPlatform\State\ProcessorInterface;
+use Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException;
+use Sylius\Bundle\ApiBundle\Exception\ProductCannotBeRemoved;
+use Sylius\Component\Core\Model\ProductInterface;
+use Webmozart\Assert\Assert;
+
+final readonly class ProductProcessor implements ProcessorInterface
+{
+    public function __construct(
+        private ProcessorInterface $persistProcessor,
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/StateProcessor/Delete/ProductProcessor.php:27]: Should not be needed if it will be only for delete operation 🖖🏻 
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml [hunk 33]
```diff
@@ -49,5 +49,14 @@
             <argument type="service" id="Sylius\Bundle\CoreBundle\SectionResolver\SectionProviderInterface" />
             <tag name="api_platform.state_provider" priority="10" />
         </service>
+
+        <service id="sylius_api.state_provider.shop.product_attributes" class="Sylius\Bundle\ApiBundle\StateProvider\ProductAttributesProvider">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/state_providers.xml:53]: The `shop` part should not be necessary, as the provider is not in the `Shop` namespace
```suggestion
        <service id="sylius_api.state_provider.product_attributes" class="Sylius\Bundle\ApiBundle\StateProvider\ProductAttributesProvider">
```

**Review by GSadee — APPROVED** [review]
